### PR TITLE
Make sure secret data is zeroized on drop and not copied around

### LINKED ISSUE #8: Make sure all objects containing secret data are zeroized on Drop
Note that in `keys.rs`, if we use `SecretKey` from `elliptic_curve` and don't actually store the secret scalar ourselves, we get the correct zeroization for free.

### PR DESCRIPTION
**Pull request by fjarri**
Fixes #8 (to the extent Rust allows, at least)

The main principles of keeping secret data secret in Rust are:
- put it in a `Box` to prevent stack allocation (which may lead to unwanted copies left on stack)
- have a `Drop` implementation that zeroizes secret data

We have secret data belonging to three kinds of foreign types:
- those for which `Zeroize` is defined, but which are not zeroized on drop by default (e.g. RustCrypto `Scalar`)
- those for which `Zeroize` is not defined (e.g. `GenericArray`)
- those for which `Zeroize` is not defined, but which are zeroized on drop (e.g. RustCrypto `SecretKey`). Note that RustCrypto `SigningKey` was not zeroized until recently - it is since `ecdsa` 0.12.2

Plus, of course, our own types, but for them at least we can define traits.

Changes in the PR:
- added the `SecretBox` struct, a wrapper that ensures that its contents follow the principles above
- added the `CanBeZeroizedOnDrop` helper trait, to generalize the behavior of all the kinds of types above
- added the `SerializableToSecretArray` trait returning a `SecretBox<GenericArray>`, as opposed to `SerializableToArray`. It will be used for objects containing secret data.
- scalars in `SecretKey` methods are wrapped in `SecretBox` on exposure
- `PartialEq` implementation removed from `SecretKey`, `SecretKeyFactory`, and `Signer` (kept for tests only)
- `SecretBox` is used for the internals of `SecretKey` and `SecretKeyFactory` (`Signer` gets the benefits automatically by using `SecretKey`)
- propagated changes to bindings (`SerializableToSecretArray` + missing equality for secret objects)
- made serialization explicit for secret objects (`to_secret_bytes()` in Python and `toSecretBytes()` in WASM)
- wrapping secret data passing in `kdf`, `DEM` and `Capsule` internally
- initialized a changelog

For reviewers: 
- am I overengineering it? All this is still not a 100% guarantee that Rust won't expose the secret data accidentally (which is kind of a bummer), and we've been using a Python implementation that left copies of secret data everywhere. 
- should `Signer` consume the secret key and not be cloneable, like the backend `SigningKey` in RustCrypto? What does it achieve?

### CODE REVIEW

**Comment by tuxxy** [comment]
> am I overengineering it? All this is still not a 100% guarantee that Rust won't expose the secret data accidentally (which is kind of a bummer), and we've been using a Python implementation that left copies of secret data everywhere.

Hard to answer wrt Python scope & rust... IMHO, I think attempting to zeroize is a solid idea. I think limiting the API surface to just the rust bits is the way to go about it. This PR looks to do that, so it's fine.
I think this line is crossed when you're trying to prevent secret information leak in Python. I think it's more of a concern to construct APIs in Python that don't lead to accidental privacy issues. Something I did notice is the implementation of `__bytes__` on `SecretKey`, which is a bit of a red flag. The likelihood of a logger, a dev, etc accidentally (or on purpose) calling `__bytes__` and leaking secret data _somehow_ is more likely than a privacy compromise from secret data in memory.

> should `Signer` consume the secret key and not be cloneable, like the backend `SigningKey` in RustCrypto? What does it achieve?

IMHO, yes. See my note above about safe API constructions for handling secret things. It's mostly just misuse prevention and will help developers know (and feel) that they're doing something weird and smelly.

> should we wrap the key seed given to `kdf()` in a `SecretBox`?

It definitely doesn't hurt.

> should we wrap the return value of `Capsule::open_original()` / `Capsule::open_reencrypted()` in a `SecretBox`?'

Same my above response, I don't think it would hurt to do it. :shrug:

**Comment by fjarri** [comment]
> Something I did notice is the implementation of `__bytes__` on `SecretKey`, which is a bit of a red flag.

Would making the serialization method explicit for secret objects (like `to_secret_bytes()`) help? This will go in line with the new `SerializableToSecretArray` trait on the Rust side.

> IMHO, yes. See my note above about safe API constructions for handling secret things. It's mostly just misuse prevention and will help developers know (and feel) that they're doing something weird and smelly.

I understand in general that that was the intention, but what specific scenario is prevented by the signer consuming the key?

**Comment by fjarri** [comment]
@tuxxy wrapped every secret data I could think of in `SecretBox`, and made secret object serialization explicit. Still not sure about the `Signer` consuming the `SecretKey`.

File: umbral-pre/src/capsule.rs [hunk 1]
```diff
@@ -168,7 +174,7 @@ impl Capsule {
         }
 
         let pub_key = receiving_sk.public_key().to_point();
-        let dh_point = &precursor * &receiving_sk.to_secret_scalar();
+        let dh_point = &precursor * receiving_sk.to_secret_scalar().as_secret();
```
Comment (cygnusv) [inline umbral-pre/src/capsule.rs:177]: Just curious: why not using `&` here on `receiving_sk`? Is it because the new `SecretBox`wrapper?

**Review by cygnusv — APPROVED** [review]
This is very far from my league as I'm completely new to Rust, but seems a sensible approach. Thanks for taking care of this!

File: umbral-pre-python/docs/index.rst [hunk 2]
```diff
@@ -30,10 +30,12 @@ API reference
 
         Returns a public key corresponding to this secret key.
 
-    .. py:method:: __bytes__() -> bytes
+    .. py:method:: to_secret_bytes() -> bytes
```
[the remark this answers is not in the document]
Comment (fjarri) [reply umbral-pre-python/docs/index.rst:33]: `to_bytes()` feels like it should just be an alias for `__bytes__()`, I wanted to emphasize the "secretness" more. In `nucypher`, I always thought that `to_bytes()` was used when there was a need for additional serialization arguments; I didn't realize it was supposed to mean that the bytes are secret.

File: umbral-pre/src/capsule.rs (hunk 1 above)
Comment (fjarri) [reply umbral-pre/src/capsule.rs:177]: Initially, `&` was applied to the whole `receiving_sk.to_secret_scalar()`, which returns a scalar itself; the arithmetic operator needs a reference. `as_secret()` gives a reference already (which means that the object itself is kept in the heap, wherever the wrapping `Box` put it), and the secret is not copied to the stack as it was before.

Comment (fjarri) [reply umbral-pre/src/capsule.rs:177]: At least, not copied to the stack in our code. The `*` implementation may still make copies, and that's something we can't prevent.

**Review by piotr-roslaniec — APPROVED** [review]
I have limited experience with crypto engineering myself, but the reasoning given by @fjarri and @tux is sound and the implementation looks solid :+1:

**Comment by fjarri** [comment]
Ok, going to merge with `Signer` still cloning the key, and open a separate issue to decide that before 1.0 - see #61

**Pull request merged**